### PR TITLE
[d3d9] Support unsupported formats in LockRect

### LINKED ISSUE #1669: [d3d9] Medieval 2 Total War produces black lines on world map
The game produces weird lines at the map with DXVK.
**Examples:**
[https://cdn.discordapp.com/attachments/561359616135397396/722564654089371738/2020-06-16-231806-screenshot.jpg](https://cdn.discordapp.com/attachments/561359616135397396/722564654089371738/2020-06-16-231806-screenshot.jpg)
[https://cdn.discordapp.com/attachments/561359616135397396/722564654995603476/2020-06-16-231759-screenshot.jpg](https://cdn.discordapp.com/attachments/561359616135397396/722564654995603476/2020-06-16-231759-screenshot.jpg)
**How it is intended to look:**
[https://cdn.discordapp.com/attachments/561359616135397396/722564773933350982/2020-06-16-233053-screenshot.jpg](https://cdn.discordapp.com/attachments/561359616135397396/722564773933350982/2020-06-16-233053-screenshot.jpg)
[https://cdn.discordapp.com/attachments/561359616135397396/722564774839189504/2020-06-16-233050-screenshot.jpg](https://cdn.discordapp.com/attachments/561359616135397396/722564774839189504/2020-06-16-233050-screenshot.jpg)

### Game log:
[https://cdn.discordapp.com/attachments/461958894306983947/722595766941646899/m2tw.log](https://cdn.discordapp.com/attachments/461958894306983947/722595766941646899/m2tw.log)

### Software information
Medieval 2 Total War. 32-bit Direct3D9 game from 2006.

### System information
- GPU:x86_64 Intel(R) Core(TM) i3-6100 CPU @ 3.70GHz GenuineIntel
- GPU: Nvidia GeForce GTX 960/PCIe/SSE2
- Driver: x11-drivers/nvidia-drivers-440.82-r3
- Wine version: Wine 5.10
- DXVK version: DXVK_170

### Apitrace file(s)
-Unfortunately, the game really doesn't want to be traced. It basically froze while I tried to trace it. I did manage to get an apitrace, but I did not get as far as the point where the bug happens.
[https://cdn.discordapp.com/attachments/461958894306983947/722593740367200326/m2tw.trace.tar.gz](https://cdn.discordapp.com/attachments/461958894306983947/722593740367200326/m2tw.trace.tar.gz)

### Log files
- [https://cdn.discordapp.com/attachments/461958894306983947/722593996567740506/medieval2_d3d9.log](https://cdn.discordapp.com/attachments/461958894306983947/722593996567740506/medieval2_d3d9.log)
- [https://cdn.discordapp.com/attachments/461958894306983947/722594215078527066/medieval2_dxgi.log](https://cdn.discordapp.com/attachments/461958894306983947/722594215078527066/medieval2_dxgi.log)

### PR DESCRIPTION
**Pull request by DadSchoorse**
Fixes #1669

### CODE REVIEW

File: src/d3d9/d3d9_device.cpp [hunk 1]
```diff
@@ -4102,12 +4102,15 @@ namespace dxvk {
 
     const Rc<DxvkBuffer> mappedBuffer = pResource->GetBuffer(Subresource);
 
-    auto formatInfo = imageFormatInfo(pResource->GetFormatMapping().FormatColor);
+    DxvkFormatInfo formatInfo = *imageFormatInfo(pResource->GetFormatMapping().FormatColor);
```
Comment (K0bin) [inline src/d3d9/d3d9_device.cpp:4105]: Check if the format mapping is valid instead with `IsValid()`. See `D3D9CommonTexture::NormalizeTextureProperties`.

File: src/d3d9/d3d9_device.cpp [hunk 2]
```diff
@@ -4102,12 +4102,15 @@ namespace dxvk {
...
+    if (formatInfo.elementSize == 0)
+      formatInfo = m_adapter->GetUnsupportedFormatInfo(pResource->Desc()->Format);
```
Comment (K0bin) [inline src/d3d9/d3d9_device.cpp:4107]: D3D9Device has a function to get an unsupported format. `UnsupportedFormat(pResource->Desc()->Format);`

**Review by K0bin — APPROVED** [review]

File: src/d3d9/d3d9_device.cpp [hunk 3]
```diff
@@ -4102,12 +4102,16 @@ namespace dxvk {
 
     const Rc<DxvkBuffer> mappedBuffer = pResource->GetBuffer(Subresource);
 
-    auto formatInfo = imageFormatInfo(pResource->GetFormatMapping().FormatColor);
+    auto& formatMapping = pResource->GetFormatMapping();
+
+    DxvkFormatInfo formatInfo = formatMapping.IsValid()
```
Comment (Joshua-Ashton) [inline src/d3d9/d3d9_device.cpp:4107]: Can we keep this a pointer and change UnsupportedThingyBlah?

**Review by Joshua-Ashton — CHANGES_REQUESTED** [review]

File: src/d3d9/d3d9_device.cpp (hunk 3 above)
Comment (DadSchoorse) [reply src/d3d9/d3d9_device.cpp]: fixed?

**Pull request merged**